tracee-rules/engine/engine.go

```diff
@@ -166,6 +189,15 @@ func (engine *Engine) consumeSources(done <-chan bool) {
 	}
 }
 
+func (engine *Engine) dispatchEvent(s types.Signature, pe regosig.ParsedEvent, event types.Event) {
+	switch {
+	case strings.Contains(reflect.TypeOf(s).String(), "rego"):
```

Comment:
See here for a quick benchmark: https://gist.github.com/simar7/7f671bf4ce4bd685d4c9346f44c1e9c1

Reply:
@danielpacak @simar7 why are we not using the type assert/switch here? I remember Daniel mentioned something the doesn't work but isn't this what we want? https://play.golang.org/p/en8jJCHWiqw

Reply:
I think our case is more like https://play.golang.org/p/wETpR604pdj

@itaysk I tried one more time using switch type and the following snippet seems to be working:

```go
func (engine *Engine) dispatchEvent(s types.Signature, pe ParsedEvent) {
	switch s.(type) {
	case *regosig.RegoSignature: // Make sure to use pointer because we use pointer receiver for the RegoSignature implementation
		engine.signatures[s] <- pe
	default:
		engine.signatures[s] <- pe.Event
	}
}
```

However, it will introduce the cycle because of the engine.ParsedEvent type. We could fix that by moving the ParsedEvent to a shared package, maybe types?

@simar7 Could you explain what was the advantage of using reflect.TypeOf(s) here?

Reply:
There is no advantage in this case. The only problem there is that we lose testability if we hardcode the type to `*regosig.RegoSignature`. 

In the test for `consumeSources` we use a fake type called `regoFakeSignature` https://github.com/aquasecurity/tracee/blob/6df0969653c4567f7e6dabca5a77b639c042fb83/tracee-rules/engine/engine_test.go#L16 - When evaluating the following code, this will result in taking the `default` path instead which is not the desired behaviour. 
```
func (engine *Engine) dispatchEvent(s types.Signature, pe ParsedEvent) {
	switch s.(type) {
	case *regosig.RegoSignature: // Note this isn't the same as regoFakeSignature
		engine.signatures[s] <- pe
	default:
		engine.signatures[s] <- pe.Event
	}
}
```

The other approach if we really wanted to use switch type assert would be to manually list each gosig as a case (since they are all defined types) and default for each regosig. So something like the following:

```func (engine *Engine) dispatchEvent(s types.Signature, pe ParsedEvent) {
	switch s.(type) {
	case *stdioOverSocket, *antiDebugging: 
		engine.signatures[s] <- pe.Event
	default:
		engine.signatures[s] <- pe
	}
}
```
This doesn't scale well as it would require a change to this function every time a new Go signature is added/used.

For the above reasons I went with a `reflect.TypeOf(s)` instead. If I've missed something or if there is a better way to do it, happy to understand and discuss.

Reply:
if the only holdback from doing it using type assertion (assuming we agree it's better), is just the test, I'd sacrifice it and not he production code. But I think this is a symptom of a bigger problem, the fact that engine is knowing which concrete signature types it's invoking is breaking the entire premise of the Signature interface. I don't have a better suggestion yet, just wanted to bring this point to attention. @simar7  @danielpacak